xds/internal/balancer/gracefulswitch/gracefulswitchbalancer.go

```diff
@@ -0,0 +1,398 @@
+/*
+ *
+ * Copyright 2022 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package gracefulswitch implements a graceful switch load balancer.
+package gracefulswitch
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/balancer/base"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/internal/grpcsync"
+	"google.golang.org/grpc/resolver"
+)
+
+var errBalancerClosed = errors.New("gracefulSwitchBalancer is closed")
+
+func newGracefulSwitchBalancer(cc balancer.ClientConn, opts balancer.BuildOptions) *gracefulSwitchBalancer {
+	return &gracefulSwitchBalancer{
+		cc:              cc,
+		bOpts:           opts,
+		scToSubBalancer: make(map[balancer.SubConn]balancer.Balancer),
+		closed:          grpcsync.NewEvent(),
+		pendingState: balancer.State{
+			ConnectivityState: connectivity.Connecting,
+			Picker:            base.NewErrPicker(balancer.ErrNoSubConnAvailable),
+		},
+	}
+}
+
+type gracefulSwitchBalancer struct {
+	bOpts balancer.BuildOptions
+	cc    balancer.ClientConn
+
+	mu               sync.Mutex
+	balancerCurrent  balancer.Balancer
+	balancerPending  balancer.Balancer
+	scToSubBalancer  map[balancer.SubConn]balancer.Balancer
+	pendingState     balancer.State
+	currentLbIsReady bool
+
+	// currentMu protects operations on the current balancer.
+	currentMu sync.Mutex
+
+	closed *grpcsync.Event
+}
+
+func (gsb *gracefulSwitchBalancer) updateState(bal balancer.Balancer, state balancer.State) {
+	if gsb.closed.HasFired() {
```

Comment:
I believe this is a subtle and tiny race (which actually isn't a problem).

1a. `Close()` is called.  Yield before doing anything though.
2a. `updateState()` is called.  Checks `gsb.closed.HasFired()` -- it has not yet.  Yield now.
1b. `Close()` executes `gsb.closed.Fire(); gsb.mu.Lock(); .... gsb.mu.Unlock()`
2b. `updateState()` resumes.  `gsb.mu.Lock()`.  Oops, we're running even though the balancer was closed.

Move this inside the lock and the firing of it inside the lock and you should be fine.

Another thing is that this check isn't necessary anyway, since you clear `balancerCurrent` and `balancerPending` in ` Close()`, so `gsb.balancerCurrentOrPending()` will return `false` anyway.